akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/AccountExample10.java

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package jdocs.akka.persistence.typed;
+
+import akka.actor.typed.Behavior;
+import akka.actor.typed.javadsl.ActorContext;
+import akka.actor.typed.javadsl.Behaviors;
+import akka.persistence.typed.PersistenceId;
+import akka.persistence.typed.javadsl.CommandHandler;
+import akka.persistence.typed.javadsl.CommandHandlerBuilder;
+import akka.persistence.typed.javadsl.EventHandler;
+import akka.persistence.typed.javadsl.EventHandlerBuilder;
+import akka.persistence.typed.javadsl.PersistentBehavior10;
+
+public class AccountExample10 extends PersistentBehavior10<AccountExample10.AccountCommand, AccountExample10.AccountEvent, AccountExample10.Account, AccountExample10.AccountSnapshot> {
+
+  interface AccountCommand {}
+  public static class CreateAccount implements AccountCommand {}
+  public static class Deposit implements AccountCommand {
+    public final double amount;
+
+    public Deposit(double amount) {
+      this.amount = amount;
+    }
+  }
+  public static class Withdraw implements AccountCommand {
+    public final double amount;
+
+    public Withdraw(double amount) {
+      this.amount = amount;
+    }
+  }
+  public static class CloseAccount implements AccountCommand {}
+
+  interface AccountEvent {}
+  public static class AccountCreated implements AccountEvent {}
+  public static class Deposited implements AccountEvent {
+    public final double amount;
+
+    Deposited(double amount) {
+      this.amount = amount;
+    }
+  }
+  public static class Withdrawn implements AccountEvent {
+    public final double amount;
+
+    Withdrawn(double amount) {
+      this.amount = amount;
+    }
+  }
+  public static class AccountClosed implements AccountEvent {}
+
+  interface Account {
+
+    Account applyEvent(AccountEvent event);
+
+    AccountSnapshot toSnapshot();
+
+    static Account fromSnapshot(AccountSnapshot snapshot) {
+      switch (snapshot.lifecycle) {
+        case EMPTY: return new EmptyAccount();
+        case OPEN: return new OpenedAccount(snapshot.balance);
+        case CLOSED: return new ClosedAccount();
+        default: throw new IllegalStateException();
+      }
+    }
+  }
+  public static class EmptyAccount implements Account {
+    @Override
+    public Account applyEvent(AccountEvent event) {
+      if (event instanceof AccountCreated)
+        return new OpenedAccount(0.0);
+      else throw new IllegalStateException();
+    }
+
+    @Override
+    public AccountSnapshot toSnapshot() {
+      return new AccountSnapshot(0.0, AccountLifecycle.EMPTY);
+    }
+  }
+
+  public static class OpenedAccount implements Account {
+    private double balance;
+
+    private final EventHandler<Account, AccountEvent> eventHandler =
+        EventHandlerBuilder.<Account, AccountEvent>builder()
+            .matchEvent(Deposited.class, evt -> {
+              setBalance(getBalance() + evt.amount);
+              return OpenedAccount.this;
```

Comment:
note that OpenedAccount is mutable and we return same instance

Reply:
I think you could omit that with another `matchEvent` overload:

```scala
  def matchEvent[E <: Event](eventClass: Class[E], f: Procedure[E]): EventHandlerBuilder[State, Event] = {
    matchEvent[E](eventClass, new BiFunction[State, E, State] {
      override def apply(state: State, event: E): State = { f(event); state }
    })
  }
```

Reply:
yes, unless it makes it possible to forget to return the new state instance when an immutable state is used.

Shouldn't this be written as:
```
.matchEvent(Deposited.class, this::deposited)
```

with a
```
private OpenedAccount deposited(Deposited evt) {
    balance += evt.amount;
    return this;
}
```

Reply:
Having both mutation and returning the state seems a bit weird to me, but then again I'd just prefer no mutation at all, so maybe I'm not the best judge.

You make a good point though. It would be easy to forget the `return` and write something like:

```
            .matchEvent(AccountClosed.class, evt -> {
                new ClosedAccount();
            })
```

especially because, without the braces, it would actually correctly return the new state

```
            .matchEvent(AccountClosed.class, evt -> new ClosedAccount())
```

so it probably is better to enforce a return value.

Reply:
also with mutable state there are cases when you want to switch to another state instance, another class to model something else in the lifecycle

fluent interfaces like `account.deposited(a).withdrawn(b)` is not a foreign concept so returning this from such a method would be ok